### ISSUE TITLE
Chaincode - SIGNIFICANT API CHANGE - Run -> Invoke

### ISSUE THREAD

**Issue by srderson** [hyperledger-archives/fabric#582]
Today, the chaincode interface defines two functions.

```
// Chaincode is the standard chaincode callback interface that the chaincode developer needs to implement.
type Chaincode interface {
    // Run method will be called during init and for every transaction
    Run(stub *ChaincodeStub, function string, args []string) ([]byte, error)
    // Query is to be used for read-only access to chaincode state
    Query(stub *ChaincodeStub, function string, args []string) ([]byte, error)
}
```

There are two items we would like to change.
1. Documentation, the CLI, and the REST API use the term `Invoke` instead of `Run`. Therefore, we would like to rename the `Run` function to `Invoke`.
2. The existing `Run` function returns a `[]byte` which is mostly unused and causes confusion. We would like to remove this from the return value.

The new Chaincode interface will be

```
// Chaincode is the standard chaincode callback interface that the chaincode developer needs to implement.
type Chaincode interface {
    // Invoke method will be called during init and for every transaction
    Invoke(stub *ChaincodeStub, function string, args []string) (error)
    // Query is to be used for read-only access to chaincode state
    Query(stub *ChaincodeStub, function string, args []string) ([]byte, error)
}
```

This is a disruptive change, but the longer we wait, the harder it will be to make these types of changes. Please provide any thoughts and comments.

**Comment by binhn** [comment]
@srderson could we deprecate Run for a while; that is, the Chaincode interface would contain both Invoke and Run?

**Comment by srderson** [comment]
@binhn Yes, will look into this. I'll see if we can use reflection to check for the existence of a particular function.

**Comment by prjayach** [comment]
@srderson @binhn @muralisrini Change looks ok to me. I just had a chat with Murali. I would like the ability to know whether an invoke transaction succeeded or failed. If the /transaction/{UUID} API provided this info, that would be great. Ideally, if a chaincode invoke returned an error and that error could be propagated back and returned as part of the /transaction/{UUID} API, it would greatly help our case in handling error paths within our smart contract logic. Thanks!

**Comment by muralisrini** [comment]
@prjayach : after talking with @binh remembered the decision about handing invoke errors. We will store 500 for error along with the transaction and the actual  error returned by the "error" parameter separately. We minimize non-deterministic behavior by treating all errors as "500". You can still query the "actual" error using transaction UUID.

**Comment by prjayach** [comment]
That works for me. I dont really need different error codes at least for now.

**Comment by kletkeman** [comment]
Well, if you are going to make a change like this, how about properly separating the three modes and then also separating Deploy from Restart? As in ...

// Chaincode is the standard chaincode callback interface that the chaincode developer needs to implement.
type Chaincode interface {
    // Deploy method will be called only when actually deployed
    Deploy(stub *ChaincodeStub, function string, args []string) ([]byte, error)
    // Restart method will be called whenever the chaincode is reloaded by a restart
    Restart(stub *ChaincodeStub, function string, args []string) ([]byte, error)
    // Invoke method will be called for write-style transactions
    Run(stub *ChaincodeStub, function string, args []string) ([]byte, error)
    // Query method provides read-only access to chaincode state
    Query(stub *ChaincodeStub, function string, args []string) ([]byte, error)
}

If this is not done then we will continue to see poorly managed state machines inside contracts with ambiguous handling at startup and restart etc ...

Also bearing in mind that there is a payload within each of  these that signifies a second level of RPC API -- e.g. invoke "createAsset({a JSON pseudo-state})" or something like that ...

**Comment by bcbrock** [comment]
Rather then eliminating the return value from the INVOKE transaction, why not specify a way to return it? I believe that eventually you will need to implement a "blocking" invoke that would be able to return this value. The value could also be returned as part of the event stream. 

It seems strange to me that we can ask a chaincode to perform an arbitrary computation, but not to return a result. This is especially true since the semantics of the chaincode query have never been formally specified. Lacking a specification, my suspicion is that a chaincode query can only return values based on the state after committing a block, whereas the execution of a chaincode can observe states that exist during the dynamic formation of a block. If so, the return value of an invoke may be the only (efficient) way to observe these intermediate states. For other types of chain codes it may also be more efficient in general to have the chaincode return a value rather than requiring one or more subsequent queries to determine what happened on the invoke.

**Comment by srderson** [comment]
Closed via PR #1012
